server: factor out MainCommon as a class, with a run method

jmarantz opened the pull request:
*Description*:
This is step 2 in [the plan to improve startup performance](https://goo.gl/YxzBaB). It makes it possible instantiate the full server stack without running the event loop, so the instantiation can be timed.  This also simplifies the code required in main(), though a legacy path is retained allowing existing alternate main()s to continue to work.

*Risk Level*: Medium: this has been extensively tested but deserves scrutiny as it is part of the startup path.

*Testing*:
//test/... including sanitizers.
Added tests for the new MainCommon instantiation, as well as the legacy main_common() where some of the work was delegated to main().

This is another step toward addressing https://github.com/envoyproxy/envoy/issues/2373

mattklein123 commented:
@dnoe can you take a pass on this?

jmarantz commented:
btw I'm iterating a little bit on some issues with main_common_test.cc across CI which I understand.  The rest of the system should be stable.

jmarantz commented:
TWIMC: I have managed to get this through CI by #if-ing out main_common_test, which works fine in all scenarios except for the coverage tests.  My hypothesis is that when MainCommon instantiates and destructing SignalAction, and that sequence alters the global signal handler state so that signals_test can no longer run properly.  In coverage test, because reasons, all the _test.cc files are linked together so we don't get a new process boundary.

I also attempted to use #ifndef ENVOY_CONFIG_COVERAGE which made my coverage tests kind of run test/run_envoy_bazel_coverage.sh locally, but they timed out and didn't give me any data, so I'm not sure if that resolved the issue.  Committing that didn't work -- CI still failed in coverage/SignalsTest the same way.  I then tried committing "#if 0" and that worked.  Now I'm trying #if !ENVOY_CONFIG_COVERAGE.

I'd like to be able to include main_common_test in coverage.  I suspect that requires diving deep into signal_action.h to understand why it isn't restoring state fully, which feels too deep for this PR.

htuch commented on source/exe/main_common.cc:
```diff
@@ -21,79 +22,111 @@
-  Server::ProdComponentFactory component_factory;
-  auto local_address = Network::Utility::getLocalAddress(options.localAddressIpVersion());
-  switch (options.mode()) {
+bool MainCommonBase::run() {
```
Should we have a better name for this? This capture historically what was in the common main, but I think logically it's about overall server setup, teardown and run. We already have a `Server::Instance`, so maybe `Server::Cli`, `Server::Launcher`, etc.?

jmarantz replied:
I'm OK renaming it but that would also lose any kind of a/b correlation in the review.  AFAIK there's no way to tell git that a file was renamed; it either figures it out or it doesn't, and mostly it doesn't :)

Can this be a follow-up?

htuch replied:
Sure. Either later or in a final pass before merging, either works.

jmarantz replied:
Thanks; I'll put this in a follow-up.

jmarantz commented on source/exe/BUILD:
```diff
@@ -100,7 +95,10 @@ envoy_cc_library(
         "//source/server/config_validation:server_lib",
-    ],
+    ] + select({
+        "//bazel:disable_signal_trace": [],
```
Great idea.  Doing so now.  Actually, why is this a compile-time decision rather than an option?  Then it would be easier to test without rebuilding the entire tree.  Same with HotRestart, though that's addressed in https://github.com/envoyproxy/envoy/pull/2568.

htuch replied:
Some of these options are compile-time as we want to squash dependencies (e.g. Google gRPC) or demonstrate that site-local replacement or build time removal of features is possible (e.g. we replace during import on the Google side). Hot restart is probably a better contender to be fully dynamic, as it doesn't introduce any deps.

There's a wider conversation happening right now in https://github.com/envoyproxy/envoy/issues/2069. Basically, we need some way to provide greater build time flexibility to allow very minimal Envoy binary deployments while also support kitchen sink builds with dynamic configuration. The former might be interesting in resource constrained embedded or container environments, the latter on a traditional fat server.

jmarantz replied:
I wasn't arguing against *any* compile-time switches, but it wasn't obvious what dependency this one had.  I was thinking that if we were compiling on Windows and the APIs didn't exist, we'd just #if them out based on platform support as opposed to config, but keep the class API above that consistent.

htuch replied:
Yeah, the `signal_trace` deps are the Backtrace library, an external dependency.

jmarantz commented on source/server/server.cc:
```diff
@@ -364,23 +358,40 @@ void InstanceImpl::run() {
+
+void InstanceImpl::terminate() {
+  {
+    std::unique_lock<std::mutex> lock(terminate_mutex_);
```
I didn't find a race, but I noticed a similar bool in ThreadLocalImpl::shutdown_ that was declared atomic.  I was then going to make shutdown_ atomic but felt like the equivalent set-if-not-already-and-run-code flow was easier to read as a mutex.

I probably need to do some more debugging of a live server to understand which structures are live in which threads.  Will investigate.

htuch replied:
I think shutdown is atomic as it's used for cross-thread communication, between main and workers. I think `InstanceImp::terminate` is only used on the main thread.

jmarantz commented on source/exe/main_common.h:
```diff
@@ -1,15 +1,65 @@
+#endif
+
+  std::unique_ptr<Envoy::OptionsImpl> options_;
+  MainCommonBase base_;
```
I needed to compute the options_ before I could instantiate the Base, which takes Options& as a param for now, due to the legacy main_common.  Hopefully this can all melt away once we remove our main_common call at Google.

jmarantz commented on source/exe/main_common.cc:
```diff
@@ -21,79 +22,111 @@
+#else
+    MainCommonBase main_common(options, false);
+#endif
+    return main_common.run() ? 0 : 1;
```
Agreed.  I had an irrational desire to make the diffs less.  But that failed anyway.  Fixing.

jmarantz commented on configs/google_com_proxy_port_0.json:
```diff
@@ -0,0 +1,48 @@
+{
```
done

jmarantz commented on configs/google_com_proxy_port_0.json:
```diff
@@ -0,0 +1,48 @@
+    }],
+    "admin": {
+        "access_log_path": "/tmp/admin_access.log",
+        "address": "tcp://127.0.0.1:0"
```
I was being lazy and copied the first plausible config I found.  Done.

mattklein123 commented on configs/BUILD:
```diff
@@ -32,3 +32,8 @@ genrule(
+
+filegroup(
+    name = "google_com_proxy_port_0",
+    srcs = ["google_com_proxy_port_0.v2.yaml"],
```
In general I would like keep/evolve configs/ and examples/ as something that end users can look at and learn from. Along those lines, if this is just used for testing, can we potentially move this into the test configs directory to avoid confusion?

jmarantz replied:
actually test/config does not currently have any test configurations.  Should I make a test/configs?  Or just put it under test/exe/testdata (which would be the Google custom, more or less).

jmarantz commented on source/exe/main_common.cc:
```diff
@@ -21,79 +22,111 @@
-  }
-};
+Server::DrainManagerPtr ProdComponentFactory::createDrainManager(Server::Instance& server) {
+  return Server::DrainManagerPtr{
```
done (this code was kind of just moved, but fixed anyway).

mattklein123 commented on include/envoy/server/hot_restart.h:
```diff
@@ -78,6 +80,21 @@ class HotRestart {
+  /**
+   * @returns an allocator for stats.
+   */
+  virtual Stats::RawStatDataAllocator& stats_allocator() PURE;
```
nit: `statsAllocator()`

jmarantz replied:
done

mattklein123 commented on source/exe/main.cc:
```diff
@@ -24,32 +9,22 @@
-  try {
-    options = std::make_unique<Envoy::OptionsImpl>(argc, argv, hot_restart_version_cb,
-                                                   spdlog::level::info);
+    return main_common.run() ? 0 : 1;
```
EXIT_SUCCESS/EXIT_FAILURE here also?

mattklein123 commented on source/exe/main.cc:
```diff
@@ -24,32 +9,22 @@
+    return EXIT_SUCCESS;
   } catch (const Envoy::MalformedArgvException& e) {
-    return 1;
+    std::cerr << "MalformedArgvException: " << e.what() << std::endl;
```
What's the reasoning behind these new catch blocks? Where these things caught elsewhere and now are caught here? My main comment is that the errors we print should be user friendly so I don't think we should print the name of the exception which the user doesn't care about. Perhaps just "Envoy startup failure: ..."?

jmarantz replied:
This isn't extra; it's just moved.  It was previously on main_common.cc:92.  I removed the extra printing though, and handles the case in Server::InstanceImpl::InstanceImpl where the config fails to parse, e.g. because it is empty, which is tested in hotrestart_integration_test.  I guess an invalid json config could also cause this, or failing to open the log file.

I think it makes sense to handle EnvoyException in main.cc.  Or, we could re-cast the exception in MainCommon's ctor as one that was already handled in main.cc.  WDYT?

mattklein123 replied:
I think it's fine either way. The main thing I would watch out for is that we don't catch exceptions when the run loop starts running. I realize that wasn't really clearly codified, and feel free to add more comments, but the idea is that we can throw and exit(1) during startup, but once the main loop starts up any exceptions which aren't explicitly handled should crash the process w/ core dump. That's the way the code works today so I would just be careful that we maintain that behavior.

mattklein123 commented on source/exe/main_common.cc:
```diff
@@ -21,80 +22,112 @@
 
 #include "ares.h"
 
+#include <pthread.h>
```
nit: is this needed?

jmarantz replied:
no; was using getpid() for debugging.  removed.

mattklein123 commented on source/exe/main_common.cc:
```diff
@@ -21,80 +22,112 @@
-  } catch (Envoy::EnvoyException& e) {
-    std::cerr << "unable to initialize hot restart: " << e.what() << std::endl;
-    return 1;
+    if (hot_restart) {
```
Yeah doing the --disable-hot-restart flag will be trivial now.

mattklein123 commented on source/exe/main_common.cc:
```diff
@@ -21,80 +22,112 @@
   }
+  }
+  throw EnvoyException(fmt::format("Invalid server mode: {}", int(options_.mode())));
+  return false;
```
I don't think this can actually happen? I would just replace both lines with `NOT_REACHED`

jmarantz replied:
correct; I didn't expect it to happen.  that's just a practice I evolved for dealing with untrusted enums coming into switch statements.  NOT_REACHED sgtm.

mattklein123 commented on source/server/hot_restart_nop_impl.h:
```diff
@@ -14,6 +16,10 @@ class HotRestartNopImpl : public Server::HotRestart {
 public:
   HotRestartNopImpl(){};
 
+  Thread::BasicLockable& logLock() override { return log_lock_; }
```
nit: Move these below non-override functions, and have comment `// Server::HotRestart` above. (Easier to see overridden functions).

mattklein123 commented on source/server/server.cc:
```diff
@@ -364,23 +358,37 @@ void InstanceImpl::run() {
+  terminate();
+}
+
+void InstanceImpl::terminate() {
```
You might consider having someone at Lyft smoke test this change when ready. The shutdown stuff is tricky to get right and it's very difficult to visually review for errors.

mattklein123 commented on test/mocks/server/mocks.h:
```diff
@@ -135,6 +135,15 @@ class MockHotRestart : public HotRestart {
   MOCK_METHOD0(shutdown, void());
   MOCK_METHOD0(version, std::string());
+
+  Thread::BasicLockable& logLock() override { return log_lock_; }
```
nit: These should be normal mock definitions with ON_CALL defaults to return the members.

jmarantz replied:
I'll tackle this next.

mattklein123 commented on source/exe/main_common.cc:
```diff
@@ -21,80 +22,112 @@
+  // Hot-restart should not be specified if the support is not compiled in.
+  RELEASE_ASSERT(!hot_restart);
+#endif
+  return new OptionsImpl(argc, const_cast<char**>(argv), hot_restart_version_cb,
```
return unique_ptr here vs. raw pointer?

mattklein123 commented on source/server/hot_restart_impl.h:
```diff
@@ -120,8 +120,9 @@ class HotRestartImpl : public HotRestart,
 
-  Thread::BasicLockable& logLock() { return log_lock_; }
-  Thread::BasicLockable& accessLogLock() { return access_log_lock_; }
+  Thread::BasicLockable& logLock() override { return log_lock_; }
```
nit: Can you move these into the block under 127 below. (This is the same nit I commented on before in the other file, which I didn't read carefully enough, sorry). The idea is just to visually group the overrides so it's really easy to see which are overrides for which base class and which are free functions.

jmarantz replied:
done

tonya11en commented on source/exe/main.cc:
```diff
@@ -24,32 +9,29 @@
-      [](uint64_t max_num_stats, uint64_t max_stat_name_len) {
-        return Envoy::Server::HotRestartImpl::hotRestartVersion(max_num_stats, max_stat_name_len);
-      };
+  bool enable_hot_restart = true;
```
Not to be pedantic, but is there any reason this isn't a constexpr or const?

junr03 commented:
@jmarantz I can smoke test this at Lyft. Please let me know here that the change is ready to be plopped in a machine. If it is, I can do this tomorrow 02/13

htuch commented on include/envoy/stats/stats.h:
```diff
@@ -281,5 +281,29 @@ class StoreRoot : public Store {
+
+  /**
+   * @return RawStatData* a raw stat data block for a given stat name or nullptr if there is no more
+   *         memory available for stats. The allocator may return a reference counted data location
```
Technically, I think s/may return/should return/. We have an outstanding issue related to this at https://github.com/envoyproxy/envoy/issues/2453

htuch commented on test/exe/main_common_test.cc:
```diff
@@ -0,0 +1,76 @@
+namespace Envoy {
+
+TEST(MainCommon, ConstructDestruct) {
+  // TODO(jmarantz): I think we may need to hack the config file to pick an unused port.
```
Why isn't port zero binding sufficient?

htuch commented on test/config/integration/google_com_proxy_port_0.v2.yaml:
```diff
@@ -0,0 +1,35 @@
+    dns_lookup_family: V4_ONLY
+    lb_policy: ROUND_ROBIN
+    hosts: [{ socket_address: { address: google.com, port_value: 443 }}]
+    tls_context: { sni: www.google.com }
```
You should be able to significantly reduce this down, since you don't need any TLS or DNS lookup, HCM etc. in a server startup example. See https://github.com/envoyproxy/envoy/blob/master/test/config/integration/server_ads.yaml for a very minimal example.

dnoe commented on source/exe/main.cc:
```diff
@@ -25,31 +10,28 @@
-  return Envoy::main_common(*options);
+
+  // Run the server listener loop outside try/catch blocks, so that unexpected exceptions
+  // show up as a core-dumps for easier diagnostis.
```
typo: diagnostics

kmyerson commented:
LGTM. It'll be a bit of work to switch our integration to this new main path, but no obvious problems.

jmarantz commented:
Addressed comments from @htuch 
@kmyerson existing calls to main_common() should work, and that is at least somewhat tested in main_common_test.cc.  The intent is for us to migrate off that legacy path after integrating this, and then the legacy main_common() support can be deleted.

@junr03 can you kick off a smoke-test?  Thanks!

junr03 commented:
on it

junr03 commented:
@jmarantz I put the binary on two of our machines (staging, canary) in our edge cluster. I did a full stop/start and a hot restart, and in both cases things look ok.

jmarantz commented:
Thanks all!

htuch commented:
@junr03 @jmarantz Merged. We will roll back if there is any major breakage reported in next 48h.
